File: apps/dimi/src/tui/components/messages/shell-execution.ts
import type { Component } from '@dimi-agent/pi-tui';
import { Container, Text } from '@dimi-agent/pi-tui';

import { currentTheme } from '#/tui/theme';
import type { ToolCallBlockData, ToolResultBlockData } from '#/tui/types';

import type { ResultRenderer } from './tool-renderers/types';
import { PREVIEW_LINES } from './tool-renderers/types';
import { TruncatedOutputComponent } from './tool-renderers/truncated';

export interface ShellExecutionOptions {
  readonly command?: string;
  readonly result?: ToolResultBlockData;
  readonly expanded?: boolean;
  readonly showCommand?: boolean;
  /**
   * Max command lines to render. `undefined` means no cap — used by the
   * ctrl+o expanded view so the user can see the full multi-line command
   * even when the header preview was truncated.
   */
  readonly commandPreviewLines?: number;
  readonly resultPreviewLines?: number;
  readonly tailOutput?: boolean;
  readonly expandHint?: boolean;
}

export class ShellExecutionComponent extends Container {
  constructor(options: ShellExecutionOptions) {
    super();

    if (options.showCommand === true) {
      this.addCommandPreview(options.command ?? '', options.commandPreviewLines);
    }

    if (options.result !== undefined) {
      this.addResultPreview(
        options.result,
        options.expanded ?? false,
        options.resultPreviewLines ?? PREVIEW_LINES,
        options.tailOutput ?? false,
        options.expandHint ?? true,
      );
    }
  }

  private addCommandPreview(command: string, previewLines: number | undefined): void {
    if (command.length === 0) return;
    const allLines = command.split('\n');
    const lines = previewLines === undefined ? allLines : allLines.slice(0, previewLines);
    for (const [i, line] of lines.entries()) {
      // Distinguish the command (input) from the result (output): the `$`
      // prompt uses the dedicated shell-mode hue, the command body uses
      // `textDim`, and the result below is rendered one step dimmer in
      // `textMuted` so the two stay separable without a connecting glyph.
      const text =
        i === 0
          ? currentTheme.fg('shellMode', '$ ') + currentTheme.dim(line)
          : `  ${currentTheme.dim(line)}`;
      this.addChild(new Text(text, 2, 0));
    }
  }

  private addResultPreview(
    result: ToolResultBlockData,
    expanded: boolean,
    previewLines: number,
    tailOutput: boolean,
    expandHint: boolean,
  ): void {
    if (!result.output) return;
    this.addChild(
      new TruncatedOutputComponent(result.output, {
        expanded,
        isError: result.is_error ?? false,
        maxLines: previewLines,
        tail: tailOutput,
        expandHint,
        color: 'textMuted',
      }),
    );
  }
}

export const shellExecutionResultRenderer: ResultRenderer = (
  _toolCall: ToolCallBlockData,
  result: ToolResultBlockData,
  ctx,
): Component[] => [
  // Result only. The command preview is owned by ToolCallComponent's
  // buildCallPreview across the whole lifecycle (streaming, running, and
  // done); rendering it here too would duplicate the command once the result
  // lands.
  new ShellExecutionComponent({
    result,
    expanded: ctx.expanded,
  }),
];

File: crates/dimi-tui/src/components/messages/shell_execution.rs
//! `ShellExecutionComponent` — Bash command preview + result output
//! (port of `apps/dimi/src/tui/components/messages/shell-execution.ts`).

use crate::component::Component;
use crate::components::messages::tool_renderers::{ToolResultData, TruncatedOutputComponent};
use crate::components::text::Text;
use crate::theme::{ColorToken, current_theme};

/// Shell execution block options.
pub struct ShellExecutionOptions {
    pub command: Option<String>,
    pub result: Option<ToolResultData>,
    pub expanded: bool,
    pub show_command: bool,
    /// Max command lines to render; `None` = no cap (expanded view).
    pub command_preview_lines: Option<usize>,
    pub result_preview_lines: Option<usize>,
    pub tail_output: bool,
    pub expand_hint: bool,
}

/// Renders a `$ command` preview (optional) and the result output below it.
pub struct ShellExecutionComponent {
    children: Vec<Box<dyn Component>>,
}

impl ShellExecutionComponent {
    pub fn new(options: ShellExecutionOptions) -> Self {
        let mut children: Vec<Box<dyn Component>> = Vec::new();
        if options.show_command {
            add_command_preview(
                &mut children,
                options.command.as_deref().unwrap_or(""),
                options.command_preview_lines,
            );
        }
        if let Some(result) = options.result {
            children.push(Box::new(TruncatedOutputComponent::new(
                &result.output,
                options.expanded,
                result.is_error,
                options.result_preview_lines,
                None,
                Some(options.expand_hint),
                Some(options.tail_output),
                ColorToken::TextMuted,
            )));
        }
        ShellExecutionComponent { children }
    }
}

fn add_command_preview(
    children: &mut Vec<Box<dyn Component>>,
    command: &str,
    preview_lines: Option<usize>,
) {
    if command.is_empty() {
        return;
    }
    let all_lines: Vec<&str> = command.split('\n').collect();
    let lines: Vec<&str> = match preview_lines {
        Some(cap) => all_lines.iter().take(cap).copied().collect(),
        None => all_lines,
    };
    for (i, line) in lines.iter().enumerate() {
        // `$` prompt uses the shell-mode hue; the command body uses textDim.
        let text = if i == 0 {
            format!(
                "{}{}",
                current_theme().fg(ColorToken::ShellMode, "$ "),
                current_theme().dim(line)
            )
        } else {
            format!("  {}", current_theme().dim(line))
        };
        children.push(Box::new(Text::new(&text, 2, 0)));
    }
}

impl Component for ShellExecutionComponent {
    fn render(&mut self, width: usize) -> Vec<String> {
        let mut lines = Vec::new();
        for child in &mut self.children {
            lines.extend(child.render(width));
        }
        lines
    }

    fn invalidate(&mut self) {
        for child in &mut self.children {
            child.invalidate();
        }
    }
}
